exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/RestrictingNodeDecorator.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2020 The Exonum Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exonum.binding.core.runtime;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.exonum.binding.common.crypto.PublicKey;
+import com.exonum.binding.common.hash.HashCode;
+import com.exonum.binding.core.service.Node;
+import com.exonum.binding.core.storage.database.Snapshot;
+import com.exonum.binding.core.transaction.RawTransaction;
+import java.util.function.Function;
+
+/**
+ * Node decorator which can restrict access to the node by calling {@link #restrictAccess()} method.
+ */
+class RestrictingNodeDecorator implements Node {
```

Comment:
This class is a Node decorator, hence must adhere to its specification. Currently,
a Node is still not AutoCloseable/CloseableNativeProxy, that:
 * makes this decorator break the Node specification, by throwing unspecified exceptions
 * requires all the places that previously used Node-as-interface to use particular types:
   NodeProxy or RestrictingNodeDecorator (some even perform type casts) — ServiceRuntime, ServiceWrapper, their tests.

I suggest (as we have already briefly discussed) to redefine the Node *specification* to allow it to be _#closed_ (probably, it must be a more generic AutoCloseable, because CloseableNativeProxy is more specific), and use #close to:
 * Destroy the NodeProxy (in ServiceRuntime)
 * Prevent further access to service-specific RestrictingNodeDecorator (in ServiceWrapper)
but all — on Node.

Reply:
The node currently gets passed to `createPublicApiHandlers`, we could mention there briefly that it gets closed when the service is stopped.